src/api/app/controllers/staging/staging_projects_controller.rb

```diff
@@ -36,6 +36,18 @@ def create
   def copy
     authorize @main_project.staging
 
+    @project = Project.where(name: params[:staging_project_copy_name]).first_or_initialize
+    authorize @project, :create?
+
+    unless @project.new_record?
```

Comment:
but wasn't @dmarcoux 's point that this is true for all creates?

Reply:
It's true for all creations of staging projects. It's not like anyone would be able / authorized to do that.

Reply:
> but wasn't @dmarcoux 's point that this is true for all creates?

Yes, it is. I would put this in the policy. It is not specific to the creation of a staging project. If someone tries to create a project which already exists, he/she shouldn't be authorized to do so.

Reply:
> Edit: Actually, this is a validation. It doesn't belong in the policy.

Exactly:-)

Reply:
Could we move this validation in the job creation? So it fails whenever trying to create a copy job with a project which already exists?

Reply:
You mean like:

```ruby
def perform(staging_workflow_project_name, original_staging_project_name, staging_project_copy_name, user_id)
  project = Project.where(name: params[:staging_project_copy_name]).first_or_initialize
  raise SomeException, project.errors.full_messages.to_sentence unless project.valid?
```

and then catch it in the UI? Or do you have a service in mind?

Reply:
I wouldn't use a service for this.

Perhaps a [callback](https://guides.rubyonrails.org/v5.2/active_job_basics.html#available-callbacks), so it doesn't even reach the perform. Raising an exception in there and catching it in the controller should be fine.

Reply:
Service is the way to avoid leaking authentication and authorization to other layers. That's how the best practices say to do so.

Reply:
We already use services for similar cases in the staging project controller, like the `create` action. So I am leaning towards services, if I would have to pick one of the two.

Though I was a bit surprised to find pundit policy calls in the `StagingProjectValidator` service.